torchcde/misc.py

```diff
@@ -119,3 +119,34 @@ def validate_input_path(x, t):
                          "time dimension of size {}.".format(tuple(t.shape), t.size(0)))
 
     return t
+
+
+def torch_ffill(data):
+    """Forward fills data in a torch tensor of shape [N, L, C] along the L dim.
+
+    Args:
+        data (torch.Tensor):
+
+    Returns:
+
+    """
+    # Checks
+    assert isinstance(data, torch.Tensor)
+    assert data.dim() == 3
+
+    # Function to fill a 2d tensor
+    def fill2d(x):
+        """ Forward fills in the L dimension if L is of shape [L, N]. """
+        mask = np.isnan(x)
+        idx = np.where(~mask, np.arange(mask.shape[1]), 0)
+        np.maximum.accumulate(idx, axis=1, out=idx)
+        out = x[np.arange(idx.shape[0])[:, None], idx]
+        return out
+
+    # Reshape to [N * C, L] and fill the 2d tensor
+    N, L, C = data.size()
+    data_shaped = data.transpose(1, 2).reshape(-1, L).numpy()
+    data_fill = fill2d(data_shaped).reshape(-1, C, L)
+    data_out = torch.Tensor(data_fill).transpose(1, 2)
+
+    return data_out
```

Comment:
Untested but an alternative approach to forward-fill would be something like:
```
def forward_fill(data):
    nan_mask = torch.isnan(data)
    if nan_mask.any():
        data = data.clone()
        data_slices = iter(data.unbind(dim=-2))
        nan_mask_slices = iter(nan_mask.unbind(dim=-2))

        prev_data_slice = next(data_slices)
        next(nan_mask_slices)

        for data_slice, nan_mask_slice in zip(data_slices, nan_mask_slices):
            data_slice.masked_scatter_(nan_mask_slice, prev_data_slice.masked_select(nan_mask_slice))
            prev_data_slice = data_slice

    return data
```

This has a fast path for the common case of no NaNs, which is important. That fact aside, I don't know whether this will be meaningfull faster or slower than a torch-ified version of yours. Give it a test?

Reply:
The performs a for loop over data_slices which is bound to be slower the inbuilt torch operations. I have rewritten forward fill so it is torch functions only and added the nan_mask.any() condition